core/src/main/scala/akka/persistence/r2dbc/internal/sqlserver/SqlServerJournalDao.scala

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright (C) 2022 - 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.r2dbc.internal.sqlserver
+
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.persistence.Persistence
+import akka.persistence.r2dbc.R2dbcSettings
+import akka.persistence.r2dbc.internal.PayloadCodec.RichStatement
+import akka.persistence.r2dbc.internal.JournalDao
+import akka.persistence.r2dbc.internal.PayloadCodec
+import akka.persistence.r2dbc.internal.R2dbcExecutor
+import akka.persistence.r2dbc.internal.SerializedEventMetadata
+import akka.persistence.r2dbc.internal.Sql.Interpolation
+import akka.persistence.typed.PersistenceId
+import io.r2dbc.spi.Connection
+import io.r2dbc.spi.ConnectionFactory
+import io.r2dbc.spi.Row
+import io.r2dbc.spi.Statement
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import java.time.Instant
+import java.time.LocalDateTime
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[r2dbc] object SqlServerJournalDao {
+  private val log: Logger = LoggerFactory.getLogger(classOf[SqlServerJournalDao])
+
+  val TRUE = 1
+
+  def readMetadata(row: Row): Option[SerializedEventMetadata] = {
+    row.get("meta_payload", classOf[Array[Byte]]) match {
+      case null => None
+      case metaPayload =>
+        Some(
+          SerializedEventMetadata(
+            serId = row.get[Integer]("meta_ser_id", classOf[Integer]),
+            serManifest = row.get("meta_ser_manifest", classOf[String]),
+            metaPayload))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * Class for doing db interaction outside of an actor to avoid mistakes in future callbacks
+ */
+@InternalApi
+private[r2dbc] class SqlServerJournalDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+    ec: ExecutionContext,
+    system: ActorSystem[_])
+    extends JournalDao {
```

Comment:
Yes, I can look into that.

Reply:
> Can't we extend the Postgres Dao and only override the bare minimum

@patriknw I took your comment serious and added a big refactoring to demonstrate how we could separate the actual sql query definition+binding (which is dialect specific) from the logic+error handling (which is not dialect specific and which should be reused as proposed). This is currently only done for `SqlServerDurableStateDao` and `SqlServerJournalDao` which I recommend to look at first as they show the real benefit of this change.

My last commit to this pr is rather big and is only a proposal at this point, but it shows how this could look like. If you agree with this direction, I would probably take a step back and create a separate pr from the current `main` and add this kind of decoupling for postgres and then rebase this pr on that.

Reply:
I see how that looks more structured by having clear definitions of all sql and bind methods together. On the other hand I'm worried that this becomes over-designed and just adds more ceremony and makes it more difficult to understand via the additional indirection. It's like we have two DAO layers?

What is the benefit compared to placing the sql and bind methods in the Dao itself? I can understand that it might feel wrong to use PostgresDao as the base, so we could extract that (all common parts) to a BaseDao.

Reply:
Extracting common parts to a BaseDao would make sense to me if we agree that the overhead (of separating the sql and bind methods from the actual logic and error handling) is worth it. 

Maybe adding a new dialect is not worth the complexity / overhead it seems to introduce? I definitely do not want to open up a can of worms without getting enough value back here, so please share you thoughts.

> What is the benefit compared to placing the sql and bind methods in the Dao itself?

I actually also prefer the current approach, where the dao contains the sql and binding. Hover, if we keep that approach, it would mean I would have to duplicate lots of logic. Example is your recent [contribution](https://github.com/akka/akka-persistence-r2dbc/pull/485). Without the refactoring I would have to basically copy what you have added / changed in that pr for postgres and re-implement it in the sqlserver dialect.

Reply:
After looking and thinking about it for a bit I lean towards preferring (reasonable) duplication with extending the PostgresDAO (and possibly making more things there possible to override) over the extra clever abstraction for now. Once/if we add two-three more dialects we can reconsider.

Reply:
> and possibly making more things there possible to override

This is key to the pragmatic approach that I'm advocating. We evolve it and extract pieces and methods to override as we need, but not before we need it.

The way I would like to see it is that we extend PostgresDao and override only the things that are different for SqlServer. Still with sql in the Dao, and bindings in the Dao, where the bindings could be separate methods to be overridden when it's needed. That approach was quite successful when adding the H2 dialect, and I don't see that SqlServer would be so much different.

Then moving most of PostgresDao to BaseDao is pretty much just a naming cleanup that we can make separately afterwards.

Reply:
Sounds good. I may start a new PR based on the learnings from this one.

Reply:
Thanks, don't hesitate to share a small early draft to make sure we are aligned.